### PR TITLE
qutebrowser: 2.1.1 -> 2.2.0

### PR DESCRIPTION
**Pull request by svrana**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/browsers/qutebrowser/default.nix [hunk 1]
```diff
@@ -69,12 +69,6 @@ in mkDerivationWith python3Packages.buildPythonApplication rec {
 
   patches = [
     ./fix-restart.patch
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/browsers/qutebrowser/default.nix:72]: Please remove fetchpatch from inputs.

**Review by dotlambda — CHANGES_REQUESTED** [review]
Can we also update pdfjs?

**Comment by dotlambda** [comment]
I get
```
WARNING: QtWebEngine version mismatch - unexpected behavior might occur, please open a bug about this.
  Early version: QtWebEngine 5.15.4, Chromium 87.0.4280.144 (from PyQt)
  Real version:  QtWebEngine 5.15.3, Chromium 87.0.4280.144
```
I'm sure @The-Compiler knows what that's about.

**Comment by The-Compiler** [comment]
@dotlambda What happens there is that qutebrowser needs to know the exact QtWebEngine version for various things it does (darkmode arguments, workarounds for QtWebEngine/Chromium bugs, etc.). Unfortunately, there's no version getter in QtWebEngine, so it needs to resort to other means of finding out. When QtWebEngine is fully initialized it can parse the version from the user agent it gets from QtWebEngine, but for most workarounds, we need to know the version before that's possible.

What's happened there is that qutebrowser's guessing ended up being wrong, compared to the version extracted from the user agent later. I'll improve that specific situation (there'll be no QtWebEngine 5.15.4, so qutebrowser should guess QtWebEngine 5.15.3 with PyQtWebEngine >= 5.15.4 but < 6).

However, it actually shows a deeper issue (which also explains why it works on Arch): The most reliable way for qutebrowser to get that information is to parse it from the `libQt5WebEngineCore.so` file (using a regex, I wish I was kidding, but alas, it's the only thing that works other than guessing).

I'd expect that parsing to work on NixOS, but the "(from PyQt)" indicates that it failed. Could you launch with `--debug` and show the full logs up to that warning?

**Comment by dotlambda** [comment]
It says
```
No QtWebEngine .so found in /nix/store/afk7n4smc775fw21nv1dgrk5a199csdq-qtbase-5.15.2/lib
```
That's because the correct path is
```
/nix/store/i54rflnrgibbqjmn09si163i0svvdd98-qtwebengine-5.15.3-a059e74/lib/libQt5WebEngineCore.so
```
`QLibraryInfo.LibrariesPath` just can't work with Nix. I'll add a patch.

**Comment by The-Compiler** [comment]
Happy to handle this upstream if there's a way qutebrowser can find the correct path.

**Comment by dotlambda** [comment]
> Happy to handle this upstream if there's a way qutebrowser can find the correct path.

I wouldn't know of any unless PyQtWebEngine exposes the path.
The only hacky way I could come up with is reading the `RUNPATH` of `lib/python3.8/site-packages/PyQt5/QtWebEngine.so` which will, on NixOS, list the paths to search `libQt5WebEngine.so` in. In addition, `$LD_LIBRARY_PATH` would have to be searched of course.

<details>
<summary>

Output of `readelf -d /nix/store/g23xzn6p982ixc2a8qrw8k2zlr33c522-python3.8-PyQtWebEngine-5.15.4/lib/python3.8/site-packages/PyQt5/QtWebEngine.so`:

</summary>

```
Dynamic section at offset 0xd9d0 contains 43 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libQt5WebEngine.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libQt5WebEngineCore.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libQt5Quick.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libQt5Gui.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libQt5QmlModels.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libQt5WebChannel.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libQt5Qml.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libQt5Network.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libQt5Positioning.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libQt5Core.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libGL.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000001d (RUNPATH)            Library runpath: [/nix/store/afk7n4smc775fw21nv1dgrk5a199csdq-qtbase-5.15.2/lib:/nix/store/v3xi47p44wfdgc06hw9i90nqdja7d2xn-libGL-1.3.2/lib:/nix/store/4mccd2j6x2681nyzb64bdqv0wplabj79-qtdeclarative-5.15.2/lib:/nix/store/cm9xlykpb2856rjfx481106nfajlg79d-qtlocation-5.15.2/lib:/nix/store/hlpwjy7fx4kdhb099wd7z17la2pf3fpp-qtwebchannel-5.15.2/lib:/nix/store/i54rflnrgibbqjmn09si163i0svvdd98-qtwebengine-5.15.3-a059e74/lib:/nix/store/1jn6apz0fa9h9x7rl3v6vwiymwnjznwv-glibc-2.32-40/lib:/nix/store/hsw0sq8y0a46fxc2d7krgyr6p5dy71kd-gcc-10.2.0-lib/lib]
 0x000000000000000c (INIT)               0x5000
 0x000000000000000d (FINI)               0x9a8c
 0x0000000000000019 (INIT_ARRAY)         0xe5a8
 0x000000000000001b (INIT_ARRAYSZ)       16 (bytes)
 0x000000000000001a (FINI_ARRAY)         0xe5b8
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x0000000000000004 (HASH)               0x200
 0x000000006ffffef5 (GNU_HASH)           0x680
 0x0000000000000005 (STRTAB)             0x1098
 0x0000000000000006 (SYMTAB)             0x6a8
 0x000000000000000a (STRSZ)              5471 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0xecc0
 0x0000000000000002 (PLTRELSZ)           2184 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x45f0
 0x0000000000000007 (RELA)               0x2790
 0x0000000000000008 (RELASZ)             7776 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000000000001e (FLAGS)              BIND_NOW
 0x000000006ffffffb (FLAGS_1)            Flags: NOW
 0x000000006ffffffe (VERNEED)            0x26d0
 0x000000006fffffff (VERNEEDNUM)         4
 0x000000006ffffff0 (VERSYM)             0x25f8
 0x000000006ffffff9 (RELACOUNT)          310
 0x0000000000000000 (NULL)               0x0
```

</details>

**Comment by The-Compiler** [comment]
Hmm, tricky. While qutebrowser already has [part of an ELF parser](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/misc/elf.py) for this hack, I'd rather not have to add support for parsing the dynamic section as well...

What about reading it from `/proc/<pid>/maps` instead? From a quick look it seems to show up there (since it's already imported at the point the code runs), and parsing that should be easy enough.

**Comment by dotlambda** [comment]
> What about reading it from `/proc/<pid>/maps` instead?

That should work. If you decide to implement it, let me know and I'll test it on NixOS.
But again, we can easily add a patch if it's too much work.

**Comment by The-Compiler** [comment]
Does Nix on macOS use `.so` ELF files, or does it use Mach-O? I'd assume the latter? Then that approach won't work anyways...

Not sure what to do here. Would you suggest Nix just patching the guessing based on the PyQtWebEngine version then? I'm considering doing that upstream as well, but it will break once QtWebEngine 5.15.4 is out.

On a second thought, at least QtWebEngine 5.15.4 won't come with a new Chromium version, so most likely the version difference doesn't actually matter much in practice.

**Comment by dotlambda** [comment]
> Does Nix on macOS use `.so` ELF files, or does it use Mach-O?

The latter.

> Would you suggest Nix just patching the guessing based on the PyQtWebEngine version then?

We could for example hardcode the path to the qtwebengine used to build pyqtwebengine. I'd set [`library_name`](https://github.com/qutebrowser/qutebrowser/blob/85a41de4b54cf3da6ab8c0d41222b76836238536/qutebrowser/misc/elf.py#L319) to a constant string in the patch. Or maybe setting `library_path` to the correct value is even easier.

How does the version parsing work on macOS anyway? You only search for `libQt5WebEngineCore.so*`.

**Comment by The-Compiler** [comment]
> How does the version parsing work on macOS anyway? You only search for `libQt5WebEngineCore.so*`.

It does not, the parser only understands ELF anyways. It could fallback to essentially grepping the entire file (instead of just the ELF text section) but that's 123 MB on my system, so this will take a bit of time (slowing down qutebrowser's startup).

On macOS the fallback logic is used, which is usually fine because it works with the prebuilt .dmg/.app releases. But with Nix on macOS, I'd break even if the path in `elf.py` was correct.

> We could for example hardcode the path to the qtwebengine used to build pyqtwebengine. I'd set [`library_name`](https://github.com/qutebrowser/qutebrowser/blob/85a41de4b54cf3da6ab8c0d41222b76836238536/qutebrowser/misc/elf.py#L319) to a constant string in the patch. Or maybe setting `library_path` to the correct value is even easier.

I assume that path isn't constant though? You'd then need to update it at least for every QtWebEngine package version change?

**Comment by dotlambda** [comment]
> I assume that path isn't constant though? You'd then need to update it at least for every QtWebEngine package version change?

It would change automatically. I will push the patch I suggest.

**Comment by emmanuelrosa** [comment]
What about using an optional environment variable to read the path to QtWebEngine? The Nix package would set the variable in the wrapper, and if set then qutebrowser would use it. Any other distros which face a similar situation may then use the environment variable as well.

**Comment by The-Compiler** [comment]
Ah, if you have the infrastructure in place to patch in the path automatically, I agree that's definitely the best choice. I wasn't aware of that.

@emmanuelrosa I can add a `QUTE_WEBENGINE_LIB_DIR` environment variable if @dotlambda agrees that's a good idea. Happy to do what I can (within reason :laughing:) to reduce the need for distribution-specific patches, makes things easier for everyone involved.

**Comment by dotlambda** [comment]
Rather than a variable with the path, I think that a variable with the version makes much more sense. That would work on OSes other than Linux.

**Comment by The-Compiler** [comment]
Right. The problem with that is that qutebrowser also wants to know the underlying Chromium version, so that's something you'd need to update manually with every release of QtWebEngine.

**Comment by dotlambda** [comment]
> Does Nix have a way to automatically set the version number in the wrapper though, without manual changes needed?

Yes, that's even easier than the patch.

**Comment by The-Compiler** [comment]
Alright - https://github.com/qutebrowser/qutebrowser/commit/febb921040b6670d9b1694a6ce55ae39384d1306 adds a `QUTE_QTWEBENGINE_VERSION_OVERRIDE`. If it's set incorrectly, the warning will still be logged (as the version parsed from the user agent is still the ultimate source of truth, and not affected by the override - only things that need to happen before the UA is available and thus are based on a pile of hacks are).

**Comment by dotlambda** [comment]
Thanks a lot @The-Compiler! I think we found a nice solution which will work independently from the OS.

I pushed a commit which, at the moment, adds the following to the wrapper:
```
export QUTE_QTWEBENGINE_VERSION_OVERRIDE='5.15.3-a059e74'
```
It seems like qutebrowser's version parser has no problems with the commit hash after the version.

**Comment by emmanuelrosa** [comment]
I confirm this package is working :) Thank you.
I noticed the patch is still there, though it doesn't look like it's being used.

**Comment by dotlambda** [comment]
@emmanuelrosa True, I removed it. We will have to squash the commits upon merging.

**Comment by svrana** [comment]
Let me know if I need to squash otherwise I'll assume commiter can handle that upon merge.

Thanks for handling this @dotlambda!

**Comment by The-Compiler** [comment]
> It seems like qutebrowser's version parser has no problems with the commit hash after the version.

Yep, any non-numeric suffix is [deliberately ignored](https://github.com/qutebrowser/qutebrowser/blob/v2.2.0/qutebrowser/utils/utils.py#L119-L120) at the moment. Since PyQt development version numbers also look like `5.15.3.dev2101281751` I *think* this will probably stay that way.

**Pull request merged**